File: Tests/Phase 2/Review2/dijkstra/dijkstra.swift
import Foundation

typealias EdgeMap = [String: Int]
typealias Graph = [String: EdgeMap]

func dijkstra(graph: Graph, start: String, end: String) -> ([String], Int) {
    var dijkstraTable: EdgeMap = graph.keys.reduce(into: [:]) { (result, key) in
        result[key] = Int.max
    }
    var pathRecord: [String: String?] = graph.keys.reduce(into: [:]) { (result, key) in
        result[key] = nil
    }
    
    // Init
    var visited = Set<String>()
    var queue: [(String, Int)] = [(start, 0)]
    dijkstraTable[start] = 0
    
    while !queue.isEmpty {
        queue.sort { $0.1 < $1.1 }
        let (currentNode, currentCost) = queue.removeFirst()
        
        if visited.contains(currentNode) {
            continue
        }
        visited.insert(currentNode)
        
        let neighbors = graph[currentNode]!
        for neighbor in neighbors.keys {
            let weight = neighbors[neighbor]!
            
            if currentCost + weight < dijkstraTable[neighbor]! {
                dijkstraTable[neighbor] = currentCost + weight
                pathRecord[neighbor] = currentNode
                queue.append((neighbor, currentCost + weight))
            }
        }
    }
    
    func getShortestPath(node: String?, result: inout [String]) {
        if node == nil {
            return
        } 
        result.append(node!)
        if let nextNode = pathRecord[node!] {
            getShortestPath(node: nextNode, result: &result)
        }
    }
    
    var shortestPath: [String] = []
    getShortestPath(node: end, result: &shortestPath)
    
    return (shortestPath.reversed(), dijkstraTable[end]!)
}

// Test
let graph: Graph = [
    "A": ["B": 1, "C": 1, "D": 3],
    "B": ["A": 1, "D": 2, "E": 1],
    "C": ["A": 1, "D": 1],
    "D": ["A": 2, "B": 3, "C": 1, "E": 2],
    "E": ["B": 1, "D": 2]
]
let (result, cost) = dijkstra(graph: graph, start: "A", end: "E")
print(result, cost)

File: Miscellaneous/Other/Review/Stage 7/Review 1/Part1/dijkstra/dijkstra.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// EdgeMap represents the neighbors and their corresponding weights
type EdgeMap map[string]int

// Graph represents the whole graph where each node maps to its neighbors
type Graph map[string]EdgeMap

// queue's element
type Item struct {
	node string
	cost int
}

func dijkstra(graph Graph, start string, end string) ([]string, int) {
	dijkstraTable := make(EdgeMap)
	recordPath := make(map[string]string)

	for key := range graph {
		dijkstraTable[key] = math.MaxInt
		recordPath[key] = ""
	}

	queue := []*Item{
		{
			node: start,
			cost: 0,
		},
	}
	visited := make(map[string]bool)
	dijkstraTable[start] = 0

	for len(queue) > 0 {
		sort.Slice(queue, func(i, j int) bool {
			return queue[i].cost < queue[j].cost
		})

		current := queue[0]
		queue = queue[1:]

		current_node, current_cost := current.node, current.cost

		if _, exists := visited[current.node]; exists {
			continue
		}
		visited[current.node] = true

		for neighbor, weight := range graph[current_node] {
			if current_cost+weight < dijkstraTable[neighbor] {
				dijkstraTable[neighbor] = current_cost + weight
				queue = append(queue, &Item{
					node: neighbor,
					cost: current_cost + weight,
				})
				recordPath[neighbor] = current_node
			}
		}
	}

	shortestPath := []string{}
	var getShortestPath func(node string)
	getShortestPath = func(node string) {
		if node == "" {
			return
		}
		shortestPath = append(shortestPath, node)
		getShortestPath(recordPath[node])
	}

	getShortestPath(end)

	for i, j := 0, len(shortestPath)-1; i < j; i, j = i+1, j-1 {
		shortestPath[i], shortestPath[j] = shortestPath[j], shortestPath[i]
	}

	return shortestPath, dijkstraTable[end]
}

// test
func main() {
	graph := Graph{
		"A": {"B": 1, "C": 1, "D": 3},
		"B": {"A": 1, "D": 2, "E": 1},
		"C": {"A": 1, "D": 1},
		"D": {"A": 3, "B": 2, "C": 1, "E": 2},
		"E": {"B": 1, "D": 2},
	}

	result, cost := dijkstra(graph, "A", "E")
	fmt.Println("Shortest path:", result, "Cost:", cost)
}
